### DISCUSSION TITLE
Font size and dimensions question

### DISCUSSION THREAD

**Discussion by pzinn** [KaTeX/KaTeX#2250] [General]
this is not really a feature request, more like a question.
as far as I understand, even the unit em doesn't refer to the local font size but rather the font size outside the katex code multiplied by some factor. for instance,
```
a\raisebox{0.5em}{b}c
\qquad
A_{A_{a\raisebox{0.5em}{b}c}}
```
will produce the same vertical shift irrespective of the fact that the second \raisebox was in \scriptscriptstyle.
so my question is, is there any way to produce dimensions that are related to the current local font size? more modestly, can one build a box whose size is scaled by the current font size?

**Comment by ronkok** [comment]
Size changes, such as `\Large`, or `\scriptsize`, will change the box scale.

**Comment by edemaine** [comment]
This is a feature of TeX: `em` units refer to the (containing) text font size, not dependent on the current math style.

However, the `mu` unit is the one math-mode-specific unit, which scales with the current math style.  It's 1/18 of an em in the math font.

For example:

```latex
a\raisebox{10mu}{b}c
\qquad
A_{A_{a\raisebox{10mu}{$\scriptscriptstyle b$}c}}
\\[2ex]
a\hspace{100mu}c
\\[-1ex]
{}_{a\hspace{100mu}c}
```

![image](https://user-images.githubusercontent.com/2218736/80868777-a1f79200-8c6a-11ea-8c4b-a39c96d47103.png)

Unfortunately, in actual LaTeX, `\raisebox` and `\hspace` do not seem to allow `mu` units. Apparently "TeX provides no easy way to do math mode vertical spacing that varies with current math style like mu units." [[ref](http://tug.org/texlive/Contents/live/texmf-dist/doc/latex/amsmath/amsopn.pdf)]  So beware that the above code will (currently) work in KaTeX, but not in LaTeX.

More generally (and in both LaTeX and KaTeX), you can use `\mathchoice` to do different spacing depending on which math style you're in.

**Comment by pzinn** [comment]
Great, thanks Erik!

**Comment by pzinn** [comment]
actually, I'm still unhappy about this.
let's modify your example as follows:
```
a\hspace{70mu}c
\\[-1ex]
{}_{a\hspace{100mu}c}
```
since the font size is subscript is 0.7 times the normal one, the two `hspace` should have the exact same effect, right?
except they don't: the second one is noticeably larger than the first one.

and yes, I understand this is not a KaTeX, really just a TeX question.
it is however relevant for me as I try to use KaTeX in my web application, as all units *must* be relative for everything to work smoothly.

**Comment by pzinn** [comment]
answering myself because I begin to understand the root of the problem. (maybe this is obvious to you KaTeX developers but it wasn't to me). compare
```
22222
\\
{}_{2222222}
\\
{}_{{}_{2222222222}}
```
in KaTeX and LaTeX. in KaTeX these will 3 lines will have almost exactly the same width because KaTeX literally scales fonts, AFAICT. in LaTeX they won't at all.
now the definition of mu in LaTeX (which is essentially used as is by KaTeX) makes sense in the context of the LaTeX rendering, but not in that of KaTeX rendering, where fonts are just scaled (and so mu should just be scaled by a factor of say 1, 0.7, 0.5 for normal, script, scriptscript)
my 2c

**Comment by pzinn** [comment]
incidentally is there a way of reopening this? I don't think I can do this myself.

**Comment by k4b7** [comment]
The problem is that LaTeX uses three separate fonts for rendering type/display, script, and scriptscript sizes (cm?10, cm?7, and cm?5 respectively).  The latte two fonts are flattened a bit and aren't directly scaled versions of cm?10 glyphs.  We only use a single font size, cm?10, and let the browser scale things appropriately.  We could solve this problem in a couple ways: 
- add more fonts (not ideal since it means download more)
- try to figure out how the cm?7 and cm?5 fonts are squished and try to apply that transform to our cm?10 glyphs (this is likely glyph specific and hard to determine)

**Comment by pzinn** [comment]
I *wasn't* suggesting modifying the fonts (sorry edit). I think the choice you guys made is sound. however it does not solve my original problem which is to get a proper scaled unit (something like `scale=1` in `calculateSize` in `units.js`). Using `mu`  (as suggested by @edemaine) doesn't work. should `mu` be defined differently? maybe.

**Comment by k4b7** [comment]
I was curious what LaTeX does so I used quicklatex.com to render it and the amount the `b` is raised up looks similar to the KaTeX rendering posted above.
<img width="328" alt="Screen Shot 2020-08-01 at 11 43 28 PM" src="https://user-images.githubusercontent.com/1044413/89115009-d4bf4b00-d450-11ea-81e7-2adc8fbad49e.png">
